doc/design-qemu-blockdev.rst

```diff
@@ -0,0 +1,175 @@
+to ensure compatibility with future versions fo QEMU. With this, we can also
+switch all hotplugging-related methods to use QMP and drop all code relating to
+the human monitor interface.
+
+Ganeti should support the AIO mode `io_uring` as long as the QEMU version on
+the node is recent enough.
+
+Implementation
+==============
+
+I/O Methods
++++++++++++
+
+With QEMU 5.0, support for `io_uring` has been introduced. This should be
+supported by Ganeti as well, given a recent enough QEMU version is present on
+the node. Ganeti will still default to using the `threads` mode on new
+installations.
+
+Disk Cache
+++++++++++
+
+Using the following table found in `man 1 qemu-system-x86_64` we can translate
+the disk cache modes known to Ganeti into the settings required by `-blockdev`:
+
+  .. code-block:: none
+
+    ┌─────────────┬─────────────────┬──────────────┬────────────────┐
+    │             │ cache.writeback │ cache.direct │ cache.no-flush │
+    ├─────────────┼─────────────────┼──────────────┼────────────────┤
+    │writeback    │ on              │ off          │ off            │
+    ├─────────────┼─────────────────┼──────────────┼────────────────┤
+    │none         │ on              │ on           │ off            │
+    ├─────────────┼─────────────────┼──────────────┼────────────────┤
+    │writethrough │ off             │ off          │ off            │
+    ├─────────────┼─────────────────┼──────────────┼────────────────┤
+    │directsync   │ off             │ on           │ off            │
+    ├─────────────┼─────────────────┼──────────────┼────────────────┤
+    │unsafe       │ on              │ off          │ on             │
+    └─────────────┴─────────────────┴──────────────┴────────────────┘
+
+The table also shows `directsync` and `unsafe`, which are currently not
+implemented by Ganeti and may be addressed in future changes. The Ganeti value
+of `default` should internally be mapped to `writeback`, as that reflects the
+values which QEMU assumes when not given explicitly according to the
+documentation.
+
+Construction of the command line parameters
++++++++++++++++++++++++++++++++++++++++++++
+
+The code responsible for generating the commandline parameters to add disks,
+cdroms and floppies needs to be changed to use the combination of `-blockdev`
+and `-device`. This will not change they way a disk is actually presented to
+the virtual guest.
+A small exception will be cdrom: Ganeti used to overwrite the cdrom disk type to
+`ide` when `boot_order` is set to `cdrom`. This is not required any more - QEMU
+will also boot off VirtIO or SCSI CD drives.
+
+Hot-Adding Disks
+++++++++++++++++++
+
+The code needs to be refactored to make use of the QMP method "blockdev-add",
+using the same parameters as for its commandline counterpart (e.g. blockdev node
+id, caching, aio-mode).
+
+Hot-Removing Disks
+++++++++++++++++++
+
+Hot-adding a disk consists of two steps:
```

Comment:
Shouldn't this be Hot-removing, instead of Hot-adding?

Reply:
Yes, absolutely. Thanks for spotting this!

Reply:
just tested the rbe userspace and got an error
```
cat /var/log/ganeti/kvm/test.vm.log
qemu-system-x86_64: -blockdev driver=raw,node-name=disk-222ec87a-b27d-4756,auto-read-only=off,cache.direct=off,cache.no-flush=off,discard=ignore,file.aio=threads,file.driver=host_device,file.filename=rbd:rbd/4fdf6419-ffd5-4dff-a60f-7a4274a00566.rbd.disk0: Could not open 'rbd:rbd/4fdf6419-ffd5-4dff-a60f-7a4274a00566.rbd.disk0': No such file or directory 
```